### PR TITLE
Pj/feat/cafe controller with location

### PR DESCRIPTION
**Pull request by pablisch**
Provisional controller function to use userLocation to identify suitable cafes.

- [x] function requires either facilities array and/or requiredFacilites array. If both are provided, the facilities array will be ignored.
- [x] function requires a location in the form of latitude,longitude in params, e.g. ?userLocation=51.5076,-0.0994
- [x] function takes an optional distance param as the max distance but will otherwise default, currently set to 5km.
- [x] function matches ALL requiredFacilities if provided.
- [x] function matches ANY facilities if no requiredFacilities provided.
- [x] function matches the above and cafes within given range.
- [x] distance from userLocation added to returned cafes.
- [x] Currently, cafes are ordered by distance only.

Additionally, 

- [x] the cafe Model has been changed to include location data.
- [x] seed data has had location data added and the test db has been updated.
- [x] the cafe model has a provisional addition of openingHours which is nothing to do with this feature but I noticed it was missing.

To be done later, in another ticket...
NO opening hours data has been added to seeds.
Cafes should be selected ONLY when they are open.
NOTE: This should really happen in the query and not by filtering the returned data.

### CODE REVIEW

**Comment by pablisch** [comment]
I am not intending to add seed data for opening hours at this time.
I was going to hold off this PR until user location was available but there is nothing to stop any requests from the frontend using a dummy location param as the current postman requests do. If being tested, please use any of the postman requests in the shared workspace that include location.

If this PR seems premature, please say. Personally, I feel the visibility is more helpful than keeping it hidden and it provides a clear onward path for the frontend location work to be carried out.

**Comment by siqbal181** [comment]
@pablisch Can see 11/12 tasks done - is this in DRAFT or all good to check ?

**Comment by kikilondon** [comment]
@pablisch there is still the facilityComponent error on this branch.
Screenshot 2024-03-02 at 14 38 15

**Comment by siqbal181** [comment]
> Member

@kikilondon @pablisch This is due to a change in naming from the FacilityComponent --> Facility. It will be resolved once merged into main I think.. I just changed the names in `FacilitySelectorScreen` file to reference the `Facility` component.. That or once my PR is merged into main then this one will have conflicts to resolve.. and it can just opt to keep changes from my PR?

thoughts?

File: backend/controllers/cafeController.js
Comment (siqbal181) [inline backend/controllers/cafeController.js:8]: @pablisch Do we want to have the error handling for if no facilities in BE and FE? I have a modal for the instance of not selecting a facility, which would stop it from going to the BE altogether (the required facility part is still pending I guess). Just a thought, maybe double error handling is good on both sides though

File: backend/controllers/cafeController.js
Comment (siqbal181) [inline backend/controllers/cafeController.js:18]: Console.logs not needed

File: backend/controllers/cafeController.js
Comment (siqbal181) [inline backend/controllers/cafeController.js:13]: This looks very cool! I'd be interested in a run-through or documentation of what you used to action this

File: backend/models/cafeModel.js
Comment (siqbal181) [inline backend/models/cafeModel.js:9]: maybe just add a comment here stating that it is reserved for future use

File: backend/models/cafeModel.js
Comment (siqbal181) [inline backend/models/cafeModel.js:39]: Just a q but what is this for? I get it's something needed for this location part, but interested

**Review by siqbal181 — CHANGES_REQUESTED** [review]
Looks very cool! Like I mentioned in a comment I'd love to know how you went about actioning this. Just a few comms.

File: backend/controllers/cafeController.js
Comment (pablisch) [reply backend/controllers/cafeController.js:8]: You're quite right, this should be handled in the frontend, but I think it is good practice to handle any possible errors in the backend too. Even though it is unlikely these errors will not be handled in the frontend, there is nothing to stop anyone making a manual browser request to the API causing an unhandled error and crashing the API server.

Comment (pablisch) [reply backend/controllers/cafeController.js:8]: As it goes, I handled this error really badly and the server would crash anyway. Thank you for highlighting the error handling 🙂

**Comment by pablisch** [comment]
Thanks for the feedback and questions, Sidra.
Main has been pulled into this branch so there should be no remaining issues there.
I added notes to the README regarding the sections you were asking about and more. Please ask about anything missing or just ask about anything. That was a really good reminder.

The Endpoint may only be tested in the test environment as the database holds different data currently. In the backend, run `npm run start:test`.

There are now six Postman requests in the shared workspace relating to this endpoint, all in the cafe routes folder. Each should return respectively:

- The British Library object
- An empty array
- Tate and British Library objects
- Tate object
- "message": "Both facilities and location are required."
- "message": "Both facilities and location are required."

**Review by siqbal181 — APPROVED** [review]
This is so cool Pablo! 

FYI the last postman request needs adjusting to remove the location (so it shows the error as expected)

**Comment by pablisch** [comment]
> This is so cool Pablo!
> 
> FYI the last postman request needs adjusting to remove the location (so it shows the error as expected)

Thanks, Sidra!
I forgot to press SAVE in postman!

**Pull request merged**
